Cabal autogen-modules and autogen-includes completion

rm41339 opened the pull request:
Added support for the autogen-modules: field in .cabal file completions inside library, executable, test-suite, and benchmark stanzas and autogen-includes in all relevant stanza types via libExecTestBenchCommons.

Behavior was tested manually in a cabal file.

Closes #4526

VeryMilkyJoe commented:
Thank you for the PR!
You will need to rebase your changes on top of the current master branch.
If you have trouble with the rebasing, it might be easier to reset to current master and then cherry pick your commit afterwards.
You will also need to add some tests before we can merge this, you can take some inspiration from #4535 if you want, but please do not copy paste.

rm41339 commented:
Hi, I updated the changes to reflect on top of the current master branch and added tests as well. Please let me know if there's anything else you'd like me to change!

VeryMilkyJoe commented on plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completion/Data.hs:
```diff
@@ -181,6 +181,7 @@ libExecTestBenchCommons =
       ("ghcjs-prof-options:", constantCompleter ghcOptions),
       ("ghcjs-shared-options:", constantCompleter ghcOptions),
       ("includes:", filePathCompleter),
+      ("autogen-includes:", filePathCompleter),
```
It would be better to add the `autogen-modules` completer in this spot as well, so that we don't get the duplicated code for each stanza.

VeryMilkyJoe commented on plugins/hls-cabal-plugin/test/Completer.hs:
```diff
@@ -336,6 +338,27 @@ importCompleterTests =
+    , testAutogenField "executable"      "autogen-completion/autogen-completion.cabal" (Position 11 9) ["autogen-modules:", "autogen-includes:"]
+    , testAutogenField "test-suite"      "autogen-completion/autogen-completion.cabal" (Position 16 9) ["autogen-modules:", "autogen-includes:"]
+    , testAutogenField "benchmark"       "autogen-completion/autogen-completion.cabal" (Position 21 9) ["autogen-modules:", "autogen-includes:"]
+    , testAutogenField "foreign-library" "autogen-completion/autogen-completion.cabal" (Position 26 9) ["autogen-includes:"]
```
Does this test succeed? I don't think you included the completer for foreign libraries (and neither should you have).
Did you try adding this field in a foreign-library, because from the documentation I would've assumed it is not valid there?

VeryMilkyJoe commented on plugins/hls-cabal-plugin/test/Completer.hs:
```diff
@@ -336,6 +338,27 @@ importCompleterTests =
+    , testAutogenField "test-suite"      "autogen-completion/autogen-completion.cabal" (Position 16 9) ["autogen-modules:", "autogen-includes:"]
+    , testAutogenField "benchmark"       "autogen-completion/autogen-completion.cabal" (Position 21 9) ["autogen-modules:", "autogen-includes:"]
+    , testAutogenField "foreign-library" "autogen-completion/autogen-completion.cabal" (Position 26 9) ["autogen-includes:"]
+    , testAutogenField "common"          "autogen-completion/autogen-completion.cabal" (Position 29 9) ["autogen-includes:"]
```
I don't see a reason not to check for `autogen-modules` here as well, I think we could add it.

rm41339 commented:
Hello, I've made the requested changes.

I apologize for some of the unnecessary edits, I think when I was updating to the latest version some changes got lost. I have reverted them back. 

Please let me know if there is an issue with anything else, and thank you for your patience.

rm41339 commented:
I have caught an error in my commit. My apologies, am fixing it right now. I'm still not quite sure by what you meant about adding autogen-modules to libExecTestBenchCommons?

VeryMilkyJoe commented:
Looks good to me now, if the CI is green I think we can merge :)
Just one last nitpick, I think we should rename the `testdata/autogen-completion` directory to `testdata/completion` so we can add more testdata for completions there in the future.
You already made the change, adding them both to `libExecTestBenchCommons`, so no worries, I'm happy now!

rm41339 commented:
Sounds great, thank you for checking! I've renamed the directory :)

fendor reviewed (approved):
Changes look good to me!

fendor commented:
Superseded by #4587
